File: 379-DesignPhoneDirectory.cpp
/*

Design a Phone Directory which supports the following operations:

 

get: Provide a number which is not assigned to anyone.
check: Check if a number is available or not.
release: Recycle or release a number.
Example:

// Init a phone directory containing a total of 3 numbers: 0, 1, and 2.
PhoneDirectory directory = new PhoneDirectory(3);

// It can return any available phone number. Here we assume it returns 0.
directory.get();

// Assume it returns 1.
directory.get();

// The number 2 is available, so return true.
directory.check(2);

// It returns 2, the only number that is left.
directory.get();

// The number 2 is no longer available, so return false.
directory.check(2);

// Release number 2 back to the pool.
directory.release(2);

// Number 2 is available again, return true.
directory.check(2);

*/

// init:     Time: O(n), Space: O(n)
// get:      Time: O(1), Space: O(1)
// check:    Time: O(1), Space: O(1)
// release:  Time: O(1), Space: O(1)

class PhoneDirectory {
public:
    /** Initialize your data structure here
        @param maxNumbers - The maximum numbers that can be stored in the phone directory. */
    PhoneDirectory(int maxNumbers) :
        curr_{0}, numbers_(maxNumbers), used_(maxNumbers) {  // Time: O(n), Space: O(n)

        iota(numbers_.begin(), numbers_.end(), 0);
    }
    
    /** Provide a number which is not assigned to anyone.
        @return - Return an available number. Return -1 if none is available. */
    int get() {  // Time: O(1), Space: O(1)
        if (curr_ == numbers_.size()) {
            return -1;
        }
        const auto number = numbers_[curr_++];
        used_[number] = true;
        return number;
    }
    
    /** Check if a number is available or not. */
    bool check(int number) {  // Time: O(1), Space: O(1)
        if (number < 0 || number >= numbers_.size()) {
            return false;
        }
        return !used_[number];
    }
    
    /** Recycle or release a number. */
    void release(int number) {  // Time: O(1), Space: O(1)
        if (number < 0 || number >= numbers_.size() || !used_[number]) {
            return;
        }
        used_[number] = false;
        numbers_[--curr_] = number ;
    }

private:
    int curr_;
    vector<int> numbers_;
    vector<bool> used_;
};

/**
 * Your PhoneDirectory object will be instantiated and called as such:
 * PhoneDirectory obj = new PhoneDirectory(maxNumbers);
 * int param_1 = obj.get();
 * bool param_2 = obj.check(number);
 * obj.release(number);
 */

File: java/379-DesignPhoneDirectory.java
/*

Design a Phone Directory which supports the following operations:

 

get: Provide a number which is not assigned to anyone.
check: Check if a number is available or not.
release: Recycle or release a number.
Example:

// Init a phone directory containing a total of 3 numbers: 0, 1, and 2.
PhoneDirectory directory = new PhoneDirectory(3);

// It can return any available phone number. Here we assume it returns 0.
directory.get();

// Assume it returns 1.
directory.get();

// The number 2 is available, so return true.
directory.check(2);

// It returns 2, the only number that is left.
directory.get();

// The number 2 is no longer available, so return false.
directory.check(2);

// Release number 2 back to the pool.
directory.release(2);

// Number 2 is available again, return true.
directory.check(2);

*/

/*

using 2 sets used and free: not a good idea, as it doesn't reduce 
the capacity when elements are deleted, so it gets more and more 
sparse and thus takes more and more time to find anything.

*/

class PhoneDirectory {
   boolean[] numbers;
    LinkedList<Integer> available;
    /** Initialize your data structure here
        @param maxNumbers - The maximum numbers that can be stored in the phone directory. */
    public PhoneDirectory(int maxNumbers) {
        numbers=new boolean[maxNumbers];
        available = new LinkedList<>();
        for(int i=0; i<maxNumbers; i++)
            available.add(i);
    }
    
    /** Provide a number which is not assigned to anyone.
        @return - Return an available number. Return -1 if none is available. */
    public int get() {
        if(available.isEmpty())
            return -1;
        int top =available.poll();
        numbers[top]=true;
        return top;
    }
    
    /** Check if a number is available or not. */
    public boolean check(int number) {
         if(number<0 || number>=numbers.length)
             return false;
        return numbers[number]==false;
    }
    
    /** Recycle or release a number. */
    public void release(int number) {
        if(number<0 || number>=numbers.length||numbers[number]==false)
            return;
        numbers[number]=false;
        available.add(number);
    }
}

class PhoneDirectory{

    Set<Integer> used = new HashSet<Integer>();
    Queue<Integer> available = new LinkedList<Integer>();
    int max;
    public PhoneDirectory(int maxNumbers) {
        max = maxNumbers;
        for (int i = 0; i < maxNumbers; i++) {
            available.offer(i);
        }
    }
    
    public int get() {
        Integer ret = available.poll();
        if (ret == null) {
            return -1;
        }
        used.add(ret);
        return ret;
    }
    
    public boolean check(int number) {
        if (number >= max || number < 0) {
            return false;
        }
        return !used.contains(number);
    }
    
    public void release(int number) {
        if (used.remove(number)) {
            available.offer(number);
        }
    }

}
